pkg/cloudprovider/openstack.go

```diff
@@ -0,0 +1,728 @@
+		IFAddr: ifAddr{
+			IPv4: ipv4,
+			IPv6: ipv6,
+		},
+		Capacity: capacity{
+			IPv4: ipv4Cap - ipv4UsedIPs,
+			IPv6: ipv6Cap - ipv6UsedIPs,
+		},
+	}, nil
+}
+
+// getIPsOnPort returns the number of unique IP addresses on the given port.
+// Those include both IPs in the list of FixedIP addresses and in the list of
+// allowed_address_pairs. This method is a helper for getNeutronPortNodeEgressIPConfiguration
+// and as such does not test if multiple networks of the same address family are assigned to
+// the port, given that the calling method will already have done this.
+func (o *OpenStack) getIPsOnPort(p neutronports.Port) (int, int) {
+	ipv4UsedIPs := make(map[string]struct{})
+	ipv6UsedIPs := make(map[string]struct{})
+
+	for _, ip := range p.FixedIPs {
+		if utilnet.IsIPv4(net.ParseIP(ip.IPAddress)) {
+			ipv4UsedIPs[ip.IPAddress] = struct{}{}
+		} else {
+			ipv6UsedIPs[ip.IPAddress] = struct{}{}
+		}
+	}
+	for _, ip := range p.AllowedAddressPairs {
+		if utilnet.IsIPv4(net.ParseIP(ip.IPAddress)) {
+			ipv4UsedIPs[ip.IPAddress] = struct{}{}
+		} else {
+			ipv6UsedIPs[ip.IPAddress] = struct{}{}
+		}
+	}
+
+	return len(ipv4UsedIPs), len(ipv6UsedIPs)
+}
+
+// getPortQuota returns the current project's quota for the maximum number of
+// ports that can be created.
+// A quota of -1 in neutron means "unlimited", hence if the quota is < 0,
+// set it to the
+func (o *OpenStack) getPortQuota() (int, error) {
+	quotas, err := neutronquotas.Get(o.neutronClient, o.projectID).Extract()
+	if err != nil {
+		return 0, err
+	}
+	if quotas.Port < 0 || quotas.Port > openstackMaxCapacity {
+		return openstackMaxCapacity, nil
+	}
+	return quotas.Port, nil
+}
+
+// reserveNeutronIPAddress creates a new unattached neutron port with the given IP on
+// the given subnet. This will server as our IPAM as it is impossible to create 2 ports
+// with the same IP on the same subnet. The created port will be identified with a custom
+// json encoded description field.
+// NOTE: We are not using tags. According to the neutron API, it's possible to add a tag when creating
+// a port. But gophercloud does not allow us to do that and we must use a 2 step process (create port, then
+// add tag). Therefore, putting what we need into the Description field allows for atomic port creation and
+// for more flexibility thanks to the JSON encoded message.
+func (o *OpenStack) reserveNeutronIPAddress(s neutronsubnets.Subnet, ip net.IP, nodeName string) (*neutronports.Port, error) {
+	opts := neutronports.CreateOpts{
+		NetworkID: s.NetworkID,
+		FixedIPs: []neutronports.IP{
+			{
+				SubnetID:  s.ID,
+				IPAddress: ip.String(),
+			},
+		},
+	}
+
+	var d []byte
+	description := EgressIPDescription{
+		Type:     egressIPTag,
+		NodeName: nodeName,
+	}
+	d, err := json.Marshal(description)
+	if err != nil {
+		return nil, fmt.Errorf("Could not marshal object %v into JSON for the IP address reservation process", description)
+	}
+	opts.Description = string(d)
```

Comment:
silly NIT: If you set variable `d` first, you could just pass in `Description: string(d)` in neutronports.CreateOpts{

Reply:
done :-)